File: panel/src/lib/api/usage.ts
import api from "./client";
import { isMockMode } from "@/lib/mock-data";
import type {
  UsageSummary,
  AgentUsageRow,
  TeamUsageRow,
  ModelUsageSlice,
  UsageTimePoint,
  UsageProjection,
  CacheEfficiencyResponse,
  UsageSession,
} from "@/types";

export type UsagePeriod = "24h" | "7d" | "30d";

// =============================================================================
// MOCK DATA  — shapes must exactly match the real backend response schemas
// =============================================================================

function mockSummary(period: UsagePeriod = "24h"): UsageSummary {
  const scale = period === "30d" ? 30 : period === "7d" ? 7 : 1;
  const base = 124_800 * scale;
  return {
    tokens_input: Math.round(base * 0.55),
    tokens_output: Math.round(base * 0.35),
    total_tokens: base,
    total_cost_usd: parseFloat((base * 0.000030).toFixed(6)),
    trend_pct: 12.5,
    period,
  };
}

function mockTimeSeries(period: UsagePeriod = "24h"): UsageTimePoint[] {
  const now = new Date();
  const points = period === "24h" ? 24 : period === "7d" ? 7 : 30;
  const step = period === "24h" ? "hour" : "day";
  return Array.from({ length: points }, (_, i) => {
    const ts = new Date(now);
    if (step === "hour") {
      ts.setHours(now.getHours() - (points - 1 - i), 0, 0, 0);
    } else {
      ts.setDate(now.getDate() - (points - 1 - i));
      ts.setHours(0, 0, 0, 0);
    }
    const base = 3_000 + Math.round(Math.random() * 4_000);
    const tokens_input = Math.round(base * 0.55);
    const tokens_output = Math.round(base * 0.35);
    const total_tokens = base;
    return {
      bucket: ts.toISOString(),
      tokens_input,
      tokens_output,
      total_tokens,
      cost_usd: parseFloat((total_tokens * 0.000030).toFixed(6)),
    };
  });
}

function mockAgentUsage(period: UsagePeriod = "24h"): AgentUsageRow[] {
  const scale = period === "30d" ? 30 : period === "7d" ? 7 : 1;
  const agents = [
    { agent_slug: "be-dev-1" },
    { agent_slug: "be-dev-2" },
    { agent_slug: "fe-dev-1" },
    { agent_slug: "fe-dev-2" },
    { agent_slug: "ux-dev-1" },
    { agent_slug: "be-qa" },
    { agent_slug: "fe-qa" },
    { agent_slug: "main-pm" },
  ];
  const grand = agents.length * 15_000 * scale;
  return agents.map((a) => {
    const ti = Math.round((5_000 + Math.random() * 20_000) * scale);
    const to_ = Math.round(ti * 0.65);
    const total = ti + to_;
    return {
      agent_slug: a.agent_slug,
      tokens_input: ti,
      tokens_output: to_,
      total_tokens: total,
      cost_usd: parseFloat((total * 0.000030).toFixed(6)),
      pct_of_total: parseFloat(((total / grand) * 100).toFixed(2)),
    };
  });
}

function mockTeamUsage(period: UsagePeriod = "24h"): TeamUsageRow[] {
  const scale = period === "30d" ? 30 : period === "7d" ? 7 : 1;
  const teams = ["backend", "frontend", "ux_ui", "main_pm"];
  const grand = teams.length * 50_000 * scale;
  return teams.map((team) => {
    const ti = Math.round((30_000 + Math.random() * 40_000) * scale);
    const to_ = Math.round(ti * 0.65);
    const total = ti + to_;
    return {
      team,
      tokens_input: ti,
      tokens_output: to_,
      total_tokens: total,
      cost_usd: parseFloat((total * 0.000030).toFixed(6)),
      pct_of_total: parseFloat(((total / grand) * 100).toFixed(2)),
    };
  });
}

function mockModelUsage(period: UsagePeriod = "24h"): ModelUsageSlice[] {
  const scale = period === "30d" ? 30 : period === "7d" ? 7 : 1;
  const models = [
    { model: "claude-opus-4", share: 0.548 },
    { model: "claude-sonnet-4", share: 0.346 },
    { model: "claude-haiku-4", share: 0.106 },
  ];
  const base = 124_800 * scale;
  return models.map((m) => {
    const ti = Math.round(base * m.share * 0.55);
    const to_ = Math.round(base * m.share * 0.35);
    const total = Math.round(base * m.share);
    return {
      model: m.model,
      tokens_input: ti,
      tokens_output: to_,
      total_tokens: total,
      cost_usd: parseFloat((total * 0.000030).toFixed(6)),
      pct_of_total: parseFloat((m.share * 100).toFixed(1)),
    };
  });
}

function mockProjection(): UsageProjection {
  const total_cost_7d = parseFloat((124_800 * 7 * 0.000030).toFixed(6));
  return {
    total_cost_7d,
    avg_daily_cost_usd: parseFloat((total_cost_7d / 7).toFixed(6)),
    projected_monthly_cost_usd: parseFloat((total_cost_7d / 7 * 30).toFixed(4)),
    basis_days: 7,
  };
}

function mockCacheEfficiency(period: UsagePeriod = "24h"): CacheEfficiencyResponse {
  return {
    cache_hit_rate: 0.3142,
    tokens_cache_read: 39_168,
    tokens_cache_write: 12_480,
    tokens_input: 85_632,
    cost_saved_by_cache_usd: parseFloat((39_168 * (3.00 - 0.30) / 1_000_000).toFixed(6)),
    period,
  };
}

function mockSessions(): UsageSession[] {
  const models = ["claude-opus-4", "claude-sonnet-4", "claude-haiku-4"];
  const agentSlugs = ["be-dev-1", "be-dev-2", "fe-dev-1", "fe-qa", "main-pm"];
  return Array.from({ length: 35 }, (_, i) => {
    const agent_slug = agentSlugs[i % agentSlugs.length];
    const model = models[i % models.length];
    const input = Math.round(2_000 + Math.random() * 8_000);
    const output = Math.round(500 + Math.random() * 3_000);
    const cache = Math.round(100 + Math.random() * 1_000);
    const started = new Date(Date.now() - (i + 1) * 12 * 60_000);
    const ended = i < 3 ? null : new Date(started.getTime() + Math.round(5 + Math.random() * 55) * 60_000);
    return {
      id: `session-mock-${i + 1}`,
      agent_slug,
      started_at: started.toISOString(),
      ended_at: ended ? ended.toISOString() : null,
      tokens_input: input,
      tokens_output: output,
      tokens_cache: cache,
      total_tokens: input + output + cache,
      cost: parseFloat(((input + output) * 0.00003 + cache * 0.000003).toFixed(4)),
      model,
    };
  });
}

// =============================================================================
// API OBJECT
// =============================================================================

export const usageApi = {
  /** Aggregated token usage summary — GET /usage/summary?period= */
  getUsageSummary: async (period: UsagePeriod = "24h"): Promise<UsageSummary> => {
    if (isMockMode()) return mockSummary(period);
    const { data } = await api.get<UsageSummary>("/usage/summary", {
      params: { period },
    });
    return data;
  },

  /** Bucketed time-series — GET /usage/time-series?period= */
  getUsageTimeSeries: async (period: UsagePeriod = "24h"): Promise<UsageTimePoint[]> => {
    if (isMockMode()) return mockTimeSeries(period);
    const { data } = await api.get<UsageTimePoint[]>("/usage/time-series", {
      params: { period },
    });
    return data;
  },

  /** Per-agent usage rows — GET /usage/by-agent?period= */
  getAgentUsage: async (period: UsagePeriod = "24h"): Promise<AgentUsageRow[]> => {
    if (isMockMode()) return mockAgentUsage(period);
    const { data } = await api.get<AgentUsageRow[]>("/usage/by-agent", {
      params: { period },
    });
    return data;
  },

  /** Per-team usage rows — GET /usage/by-team?period= */
  getTeamUsage: async (period: UsagePeriod = "24h"): Promise<TeamUsageRow[]> => {
    if (isMockMode()) return mockTeamUsage(period);
    const { data } = await api.get<TeamUsageRow[]>("/usage/by-team", {
      params: { period },
    });
    return data;
  },

  /** Per-model usage slices — GET /usage/by-model?period= */
  getModelUsage: async (period: UsagePeriod = "24h"): Promise<ModelUsageSlice[]> => {
    if (isMockMode()) return mockModelUsage(period);
    const { data } = await api.get<ModelUsageSlice[]>("/usage/by-model", {
      params: { period },
    });
    return data;
  },

  /** Monthly cost projection — GET /usage/projection */
  getUsageProjection: async (): Promise<UsageProjection> => {
    if (isMockMode()) return mockProjection();
    const { data } = await api.get<UsageProjection>("/usage/projection");
    return data;
  },

  /** Cache efficiency stats — GET /usage/cache-efficiency?period= */
  getCacheEfficiency: async (period: UsagePeriod = "24h"): Promise<CacheEfficiencyResponse> => {
    if (isMockMode()) return mockCacheEfficiency(period);
    const { data } = await api.get<CacheEfficiencyResponse>("/usage/cache-efficiency", {
      params: { period },
    });
    return data;
  },

  /**
   * Recent inference sessions — mock-mode only.
   *
   * The backend has no /usage/sessions endpoint.  In production this
   * returns an empty array so SessionsTable shows a graceful "no data"
   * state instead of throwing a 404.
   */
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  getUsageSessions: async (_limit: number = 100): Promise<UsageSession[]> => {
    if (isMockMode()) return mockSessions();
    return [];
  },
};

File: roboco/services/usage.py
"""
Usage Analytics Service

Provides token usage analytics over agent_spawn_sessions and
daily_usage_rollups tables. Supports period-based queries (24h, 7d, 30d)
and aggregation by agent, team, and model.
"""

from __future__ import annotations

from datetime import UTC, datetime, timedelta
from typing import TYPE_CHECKING, Any

from sqlalchemy import func, select

if TYPE_CHECKING:
    from sqlalchemy.ext.asyncio import AsyncSession
    from sqlalchemy.orm import InstrumentedAttribute

from roboco.db.tables import AgentSpawnSessionTable, DailyUsageRollupTable
from roboco.services.base import BaseService


def _row_tokens(row: Any) -> tuple[int, int, int, int]:
    """Extract (input, output, cache_read, cache_write) token counts as ints.

    Centralizes the null-coalescing that would otherwise be repeated across
    every aggregation method.
    """
    return (
        int(row.tokens_input or 0),
        int(row.tokens_output or 0),
        int(row.tokens_cache_read or 0),
        int(row.tokens_cache_write or 0),
    )


def _session_row(row: Any) -> dict[str, Any]:
    """Shape one spawn-session row for the dashboard's sessions table."""
    tin, tout, tcr, tcw = _row_tokens(row)
    return {
        "id": str(row.id),
        "agent_slug": row.agent_slug,
        "model": row.model,
        "started_at": row.started_at.isoformat(),
        "ended_at": row.ended_at.isoformat() if row.ended_at else None,
        "tokens_input": tin,
        "tokens_output": tout,
        "tokens_cache": tcr + tcw,
        "total_tokens": tin + tout + tcr + tcw,
        "cost": float(row.estimated_cost_usd or 0.0),
    }


def _parse_period(period: str) -> tuple[datetime, int]:
    """Parse period string into (start_dt, hours).

    Accepts '24h', '7d', '30d'. Defaults to 24h for unknown values.
    Returns (start_datetime_utc, total_hours).
    """
    now = datetime.now(UTC)
    if period == "7d":
        return now - timedelta(days=7), 7 * 24
    if period == "30d":
        return now - timedelta(days=30), 30 * 24
    # default 24h
    return now - timedelta(hours=24), 24


class UsageService(BaseService):
    """Analytics service for token usage data."""

    # =========================================================================
    # SUMMARY
    # =========================================================================

    async def get_summary(self, period: str = "24h") -> dict[str, Any]:
        """Return aggregated token and cost totals for the given period.

        Queries daily_usage_rollups for whole-day periods; falls back to
        agent_spawn_sessions for sub-day precision.

        Returns dict with: tokens_input, tokens_output, total_tokens,
        total_cost_usd, trend_pct.
        """
        start_dt, hours = _parse_period(period)

        # Current period totals from closed sessions
        result = await self.session.execute(
            select(
                func.coalesce(func.sum(AgentSpawnSessionTable.tokens_input), 0).label(
                    "tokens_input"
                ),
                func.coalesce(func.sum(AgentSpawnSessionTable.tokens_output), 0).label(
                    "tokens_output"
                ),
                func.coalesce(
                    func.sum(AgentSpawnSessionTable.tokens_cache_read), 0
                ).label("tokens_cache_read"),
                func.coalesce(
                    func.sum(AgentSpawnSessionTable.tokens_cache_write), 0
                ).label("tokens_cache_write"),
                func.coalesce(
                    func.sum(AgentSpawnSessionTable.estimated_cost_usd), 0.0
                ).label("total_cost_usd"),
            ).where(
                AgentSpawnSessionTable.started_at >= start_dt,
                AgentSpawnSessionTable.ended_at.isnot(None),
            )
        )
        row = result.one()
        tokens_input = int(row.tokens_input or 0)
        tokens_output = int(row.tokens_output or 0)
        total_cost = float(row.total_cost_usd or 0.0)
        total_tokens = (
            tokens_input
            + tokens_output
            + int(row.tokens_cache_read or 0)
            + int(row.tokens_cache_write or 0)
        )

        # Previous period for trend calculation.
        # Sum all 4 token columns so the comparison is consistent with the
        # current-period total_tokens (which also sums all 4 columns).
        prev_start = start_dt - timedelta(hours=hours)
        prev_result = await self.session.execute(
            select(
                func.coalesce(
                    func.sum(
                        AgentSpawnSessionTable.tokens_input
                        + AgentSpawnSessionTable.tokens_output
                        + AgentSpawnSessionTable.tokens_cache_read
                        + AgentSpawnSessionTable.tokens_cache_write
                    ),
                    0,
                ).label("total")
            ).where(
                AgentSpawnSessionTable.started_at >= prev_start,
                AgentSpawnSessionTable.started_at < start_dt,
                AgentSpawnSessionTable.ended_at.isnot(None),
            )
        )
        prev_row = prev_result.one()
        prev_total = int(prev_row.total or 0)

        if prev_total > 0:
            trend_pct = round((total_tokens - prev_total) / prev_total * 100, 1)
        elif total_tokens > 0:
            trend_pct = 100.0
        else:
            trend_pct = 0.0

        return {
            "tokens_input": tokens_input,
            "tokens_output": tokens_output,
            "total_tokens": total_tokens,
            "total_cost_usd": round(total_cost, 6),
            "trend_pct": trend_pct,
            "period": period,
        }

    # =========================================================================
    # TIME SERIES
    # =========================================================================

    async def get_time_series(self, period: str = "24h") -> list[dict[str, Any]]:
        """Return bucketed time-series data points.

        - 24h → hourly buckets
        - 7d / 30d → daily buckets

        Each point has: bucket (ISO string), tokens_input, tokens_output,
        total_tokens, cost_usd.

        total_tokens includes all 4 token types (input + output + cache_read +
        cache_write) so it is consistent with get_summary()'s total_tokens
        field — the two sums must match for the same period.
        """
        start_dt, _hours = _parse_period(period)

        if period == "24h":
            # Hourly buckets
            trunc_fn = func.date_trunc("hour", AgentSpawnSessionTable.started_at)
        else:
            # Daily buckets
            trunc_fn = func.date_trunc("day", AgentSpawnSessionTable.started_at)

        result = await self.session.execute(
            select(
                trunc_fn.label("bucket"),
                func.coalesce(func.sum(AgentSpawnSessionTable.tokens_input), 0).label(
                    "tokens_input"
                ),
                func.coalesce(func.sum(AgentSpawnSessionTable.tokens_output), 0).label(
                    "tokens_output"
                ),
                func.coalesce(
                    func.sum(AgentSpawnSessionTable.tokens_cache_read), 0
                ).label("tokens_cache_read"),
                func.coalesce(
                    func.sum(AgentSpawnSessionTable.tokens_cache_write), 0
                ).label("tokens_cache_write"),
                func.coalesce(
                    func.sum(AgentSpawnSessionTable.estimated_cost_usd), 0.0
                ).label("cost_usd"),
            )
            .where(
                AgentSpawnSessionTable.started_at >= start_dt,
                AgentSpawnSessionTable.ended_at.isnot(None),
            )
            .group_by(trunc_fn)
            .order_by(trunc_fn)
        )
        rows = result.fetchall()

        points = []
        for r in rows:
            ti = int(r.tokens_input or 0)
            to_ = int(r.tokens_output or 0)
            tcr = int(r.tokens_cache_read or 0)
            tcw = int(r.tokens_cache_write or 0)
            points.append(
                {
                    "bucket": r.bucket.isoformat() if r.bucket else None,
                    "tokens_input": ti,
                    "tokens_output": to_,
                    "total_tokens": ti + to_ + tcr + tcw,
                    "cost_usd": round(float(r.cost_usd or 0.0), 6),
                }
            )
        return points

    # =========================================================================
    # BY-DIMENSION (agent / team / model)
    # =========================================================================

    async def _aggregate_by(
        self,
        group_column: InstrumentedAttribute[Any],
        key_name: str,
        period: str,
    ) -> list[dict[str, Any]]:
        """Aggregate token usage grouped by an arbitrary column.

        Shared implementation behind get_by_agent/get_by_team/get_by_model.
        ``key_name`` is the dict key the grouping value is emitted under
        (e.g. "agent_slug"). Rows are ordered by input+output desc and each
        item carries pct_of_total computed against the grand total.
        """
        start_dt, _ = _parse_period(period)

        result = await self.session.execute(
            select(
                group_column.label(key_name),
                func.coalesce(func.sum(AgentSpawnSessionTable.tokens_input), 0).label(
                    "tokens_input"
                ),
                func.coalesce(func.sum(AgentSpawnSessionTable.tokens_output), 0).label(
                    "tokens_output"
                ),
                func.coalesce(
                    func.sum(AgentSpawnSessionTable.tokens_cache_read), 0
                ).label("tokens_cache_read"),
                func.coalesce(
                    func.sum(AgentSpawnSessionTable.tokens_cache_write), 0
                ).label("tokens_cache_write"),
                func.coalesce(
                    func.sum(AgentSpawnSessionTable.estimated_cost_usd), 0.0
                ).label("cost_usd"),
            )
            .where(
                AgentSpawnSessionTable.started_at >= start_dt,
                AgentSpawnSessionTable.ended_at.isnot(None),
            )
            .group_by(group_column)
            .order_by(
                func.sum(
                    AgentSpawnSessionTable.tokens_input
                    + AgentSpawnSessionTable.tokens_output
                ).desc()
            )
        )
        rows = result.fetchall()

        grand_total = sum(sum(_row_tokens(r)) for r in rows)
        items = []
        for r in rows:
            ti, to_, tcr, tcw = _row_tokens(r)
            total = ti + to_ + tcr + tcw
            items.append(
                {
                    key_name: getattr(r, key_name),
                    "tokens_input": ti,
                    "tokens_output": to_,
                    "total_tokens": total,
                    "cost_usd": round(float(r.cost_usd or 0.0), 6),
                    "pct_of_total": round(total / grand_total * 100, 2)
                    if grand_total > 0
                    else 0.0,
                }
            )
        return items

    async def get_by_agent(self, period: str = "24h") -> list[dict[str, Any]]:
        """Return per-agent token usage with pct_of_total."""
        return await self._aggregate_by(
            AgentSpawnSessionTable.agent_slug, "agent_slug", period
        )

    async def get_by_team(self, period: str = "24h") -> list[dict[str, Any]]:
        """Return per-team token usage with pct_of_total."""
        return await self._aggregate_by(AgentSpawnSessionTable.team, "team", period)

    async def get_by_model(self, period: str = "24h") -> list[dict[str, Any]]:
        """Return per-model token usage with pct_of_total."""
        return await self._aggregate_by(AgentSpawnSessionTable.model, "model", period)

    # =========================================================================
    # PROJECTION
    # =========================================================================

    async def get_projection(self) -> dict[str, Any]:
        """Return projected monthly cost based on 7-day rolling average.

        Computes the average daily cost over the last 7 days and extrapolates
        to 30 days.
        """
        seven_days_ago = datetime.now(UTC) - timedelta(days=7)

        result = await self.session.execute(
            select(
                func.coalesce(
                    func.sum(AgentSpawnSessionTable.estimated_cost_usd), 0.0
                ).label("total_cost_7d"),
                func.coalesce(func.count(AgentSpawnSessionTable.id), 0).label(
                    "session_count"
                ),
            ).where(
                AgentSpawnSessionTable.started_at >= seven_days_ago,
                AgentSpawnSessionTable.ended_at.isnot(None),
            )
        )
        row = result.one()
        total_cost_7d = float(row.total_cost_7d or 0.0)
        avg_daily_cost = total_cost_7d / 7.0
        projected_monthly = avg_daily_cost * 30.0

        return {
            "total_cost_7d": round(total_cost_7d, 6),
            "avg_daily_cost_usd": round(avg_daily_cost, 6),
            "projected_monthly_cost_usd": round(projected_monthly, 4),
            "basis_days": 7,
        }

    # =========================================================================
    # CACHE EFFICIENCY
    # =========================================================================

    async def get_cache_efficiency(self, period: str = "24h") -> dict[str, Any]:
        """Return cache hit rate and estimated savings from prompt caching.

        cache_hit_rate = cache_read_tokens / (input_tokens + cache_read_tokens)
        cost_saved = what cache reads would have cost at full input price
                     minus what they actually cost at cache-read price.
        """
        start_dt, _ = _parse_period(period)

        result = await self.session.execute(
            select(
                func.coalesce(func.sum(AgentSpawnSessionTable.tokens_input), 0).label(
                    "tokens_input"
                ),
                func.coalesce(func.sum(AgentSpawnSessionTable.tokens_output), 0).label(
                    "tokens_output"
                ),
                func.coalesce(
                    func.sum(AgentSpawnSessionTable.tokens_cache_read), 0
                ).label("tokens_cache_read"),
                func.coalesce(
                    func.sum(AgentSpawnSessionTable.tokens_cache_write), 0
                ).label("tokens_cache_write"),
            ).where(
                AgentSpawnSessionTable.started_at >= start_dt,
                AgentSpawnSessionTable.ended_at.isnot(None),
            )
        )
        row = result.one()
        tokens_input = int(row.tokens_input or 0)
        tokens_cache_read = int(row.tokens_cache_read or 0)
        tokens_cache_write = int(row.tokens_cache_write or 0)

        total_input_like = tokens_input + tokens_cache_read
        cache_hit_rate = (
            tokens_cache_read / total_input_like if total_input_like > 0 else 0.0
        )

        # Cost saved = (cache_read_tokens * full_input_rate) - actual_cache_read_cost
        # Use sonnet as the baseline (most common model) for the aggregate estimate.
        # The full input price for sonnet is $3/1M; cache read is $0.30/1M.
        _MILLION = 1_000_000.0
        _FULL_INPUT_PRICE = 3.00  # sonnet baseline, USD/1M
        _CACHE_READ_PRICE = 0.30  # 10% of input
        cost_at_full_price = tokens_cache_read * _FULL_INPUT_PRICE / _MILLION
        cost_at_cache_price = tokens_cache_read * _CACHE_READ_PRICE / _MILLION
        cost_saved = cost_at_full_price - cost_at_cache_price

        return {
            "cache_hit_rate": round(cache_hit_rate, 4),
            "tokens_cache_read": tokens_cache_read,
            "tokens_cache_write": tokens_cache_write,
            "tokens_input": tokens_input,
            "cost_saved_by_cache_usd": round(cost_saved, 6),
            "period": period,
        }

    # =========================================================================
    # TODAY'S USAGE (for CEO dashboard)
    # =========================================================================

    async def get_today_summary(self) -> dict[str, Any]:
        """Return today's aggregated usage from daily_usage_rollups.

        Used by the CEO dashboard to populate tokens_today and cost_today_usd.
        Falls back to 0 values when no data exists for today.
        """
        today = datetime.now(UTC).date()

        result = await self.session.execute(
            select(
                func.coalesce(func.sum(DailyUsageRollupTable.tokens_input), 0).label(
                    "tokens_input"
                ),
                func.coalesce(func.sum(DailyUsageRollupTable.tokens_output), 0).label(
                    "tokens_output"
                ),
                func.coalesce(
                    func.sum(DailyUsageRollupTable.tokens_cache_read), 0
                ).label("tokens_cache_read"),
                func.coalesce(
                    func.sum(DailyUsageRollupTable.tokens_cache_write), 0
                ).label("tokens_cache_write"),
                func.coalesce(
                    func.sum(DailyUsageRollupTable.total_cost_usd), 0.0
                ).label("total_cost_usd"),
            ).where(DailyUsageRollupTable.date == today)
        )
        row = result.one()

        tokens_today = (
            int(row.tokens_input or 0)
            + int(row.tokens_output or 0)
            + int(row.tokens_cache_read or 0)
            + int(row.tokens_cache_write or 0)
        )

        return {
            "tokens_today": tokens_today,
            "cost_today_usd": round(float(row.total_cost_usd or 0.0), 6),
        }

    async def get_recent_sessions(self, limit: int = 50) -> list[dict[str, Any]]:
        """Return the most recent spawn sessions, newest first.

        These are the raw per-session rows behind the aggregate panels — the
        dashboard's "Recent Sessions" table.
        """
        result = await self.session.execute(
            select(AgentSpawnSessionTable)
            .order_by(AgentSpawnSessionTable.started_at.desc())
            .limit(limit)
        )
        return [_session_row(row) for row in result.scalars().all()]


def get_usage_service(db: AsyncSession) -> UsageService:
    """Factory function matching the pattern used by other services."""
    return UsageService(db)
